### DISCUSSION TITLE
Adapting a new ( and possibly better) comic-book format.

### DISCUSSION THREAD

**Discussion by uroybd** [koreader/koreader#10442] [Ideas]
CBZ (and derivatives) have been used as an e-book format for comic books for quite a long time. While they work well on larger screens, using them on smaller handheld devices is very hard.

Many e-book reader apps solve this problem differently. Some use features like **panel zoom** and **bubble zoom**. They work, but not perfectly. Also, there's no open standard to achieve this.

This proposed second iteration of Comic Book Archive is an attempt to solve some, if possibly most, of these problems.

**I'm sharing this idea with KOReader community in the hope that KOReader can be an early adopter as it has been for new and sane ideas. I know, defining a new format is a bit of an ambitious goal, but the flexibility it will bring might be worth our effort**

## Structure
In structure, this format is similar to CBZ except for an additional `content.xml` file in the archive root. This allows these files to be **backwards compatible**.

### Content.xml
The `content.xml` file will contain several sections. A rudimentary example here:

```xml
<?xml version="1.0" encoding="utf-8"?>
<package version="2.0">
  <metadata xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:opf="http://www.idpf.org/2007/opf">
    <dc:title>Title of the comic book</dc:title>
    ... Other metadata
  </metadata>
  <content>
    <panels>
      <panel id="panel-1" href="1.png" frame="5 5 30 30" />
      <panel href="1.png" frame="40 40 30 30" />
      ...more panels
      <panel href="5.png" frame="5 5 30 100" />
      ...
    </panels>
    <bubbles>
      <bubble href="1.png" frame="5 5 30 30" parent="#panel-1" />
      <bubble href="1.png" frame="40 40 30 30" />
      ...more bubbles
      <bubble href="5.png" frame="5 5 30 100">
        Optional content in the bubble for screen reader.
      </bubble>
    </bubbles>
  </content>
</package>
```

Let's dissect the `content.xml`

#### Metadata
The metadata section contains all the metadata of the comic book. I'm using IPDF's metadata convention here.

##### Panels
This section defines all the panels in order. `href` points to the file from which we'll get the panel. `frame` contains values: x, y, width, and height.

##### Bubbles
Bubbles get defined here almost exactly like panels. Optionally, they can also have the content in them to allow the screen readers to read the comics, and with a `parent` key, a parent panel can be defined.

## Usage
Using the structure defined above, e-book reader applications can develop the following features:

### Panel Navigation
Apps can get panels from images by using the data from `content.xml` and allow users to navigate panel by panel. This allows users to read comics more easily on hand-held devices.

### Bubble Zoom
Using the `bubbles` definitions apps can enable bubble-zoom more efficiently without any image recognition burden.

Additionally, It can be used to provide better accessibility for people with disabilities affecting their reading capabilities.

**Comment by Frenzie** [comment]
It sounds interesting, though implementation might be somewhat challenging. Is there a link to a website or something for the specification and demo files?

**Reply by uroybd** [reply]
Actually no demo by far. But I think I can make one.

Also, this is an idea just born. I would like to. Feedbacks are very welcome. I'm keeping a copy of this in my digital garden and will keep it updated.

https://hermitage.utsob.me/technical-drafts/cbz-2-0-a-better-comic-book-archive/

**Reply by NiLuJe** [reply]
There's a whole digital graveyard full of attempts at doing something like that, so, err, good luck? ;o).

**Reply by uroybd** [reply]
> There's a whole digital graveyard full of attempts at doing something like that, so, err, good luck? ;o).

Can you point me to some such open ebook formats? I can find apps doing these. Google Books has bubble zoom, KOReader has panel zoom. But is there anything that is in format-level?

**Reply by NiLuJe** [reply]
I was mostly thinking of metadata-focused concerns (e.g., https://github.com/anansi-project/rfcs), but I imagine there's a similar trail of corpses for panel-view implementations...

**Reply by uroybd** [reply]
> I was mostly thinking of metadata-focused concerns (e.g., https://github.com/anansi-project/rfcs), but I imagine there's a similar trail of corpses for panel-view implementations...

Well, the metadata part is the easy one, really. EPUB's metadata schema is pretty comprehensive.

There are of course panel-view implementations available. But I found all of them at the application level. Most of them fail miserably for even mildly complex panelling.

A format-based approach has two advantages:
1. It can explicitly set panel and bubble boundaries and order.
2. The apps can leverage point 1 to detect panels and bubbles accurately without any guesswork, image recognition or whatever things they use.

I know it is too much to ask for support for a format that might not get popular at all. Yet, if I contribute in developing it, will you accept PR regarding this format?

**Reply by NiLuJe** [reply]
I feel obligated to reference our lord and savior xkcd here: https://xkcd.com/927/ ;).

Keep in mind we don't handle CBZ ourselves, we're using MuPDF. And, once you abstract the fact that we're not even currently using a modern version of it, their stance is *very* *very* much "if it's not standard, gtfo" about new formats (or even metadata schemes for existing formats, re: CBZ in particular ;p) ;).

**Reply by NiLuJe** [reply]
Case in point: find me anything that actually implements ePub 3 FXL properly? ;o).

**Comment by Frenzie** [comment]
In this particular case you could just write some Lua to do the same thing
as the column browsing modes. My main concern is that the relevant code
might gain complexity.

**Reply by uroybd** [reply]
Yes. I thought so. Should I try? It will gain complexities, true. But if it gives a better UX for the readers, it might be worth it.

**Reply by Frenzie** [reply]
A bit hard to say in advance. I'd say the more self-contained the more positive I'd be towards merging it. You could perhaps also envision it as providing an interface for a plugin to hook into, where the plugin would not necessarily be in the main repo.

Put another way, how much time you're willing to waste on a poc is up to you.  ;-)
